index.js

```diff
@@ -291,3 +292,48 @@ function getLinks (entry, decoders) {
   // raw! no links!
   return []
 }
+
+/**
+ * Create a depth-first search function.
+ * Call it with the latest links, it returns the link(s) to follow next.
+ * Maintains a queue of links it has seen but not offered up yet.
+ *
+ * In depth first, we have to resolve links one at a time; we have to
+ * find out if there are child links to follow before trying siblings
+ *
+ * e.g.
+ *
+ * o
+ * ├── x
+ * │   ├── x1
+ * │   └── x2
+ * ├── y
+ * └── z
+ *     └── z1
+ *
+ * [x, y, z] => [x]   (queue: [y, z])
+ *  [x1, x2] => [x1]  (queue: [x2, y, z])
+ *        [] => [x2]  (queue: [y, z])
+ *        [] => [y]   (queue: [z])
+ *        [] => [z]   (queue: [])
+ *      [z1] => [z1]  (queue: [])
+ */
+export function depthFirst () {
+  /** @type {import('multiformats').UnknownLink[]} */
+  let queue = []
+
+  /** @param {import('multiformats').UnknownLink[]} links */
+  return (links = []) => {
+    queue = links.concat(queue)
+    const next = queue.shift()
+    return next ? [next] : []
```

Comment:
Perhaps we can optimise this and return multiple when the links are all `raw`? i.e. we know we're at the bottom of the tree and there's no further traversal. At the moment we'll slowly tick through this 1 block at a time.

You'd need to re-add the `parallelMap` because it would matter for DFS again if it does return multiple.